src/position/position_utils.cairo

```diff
@@ -297,34 +537,170 @@ fn will_position_collateral_be_sufficient(
     market: Market,
     prices: MarketPrices,
     collateral_token: ContractAddress,
+    is_long: bool,
     values: WillPositionCollateralBeSufficientValues,
-) -> (bool, i128) {
-    // TODO
-    (true, 0)
+) -> (bool, u128) {
+    let collateral_token_price = market_utils::get_cached_token_price(
+        collateral_token, market, @prices
+    );
+    let mut remaining_collateral_usd = values.position_collateral_amount
+        * collateral_token_price.min;
+    // deduct realized pnl if it is negative since this would be paid from
+    // the position's collateral
+    if values.realized_pnl_usd < 0 {
+        remaining_collateral_usd = remaining_collateral_usd + values.realized_pnl_usd;
+    }
+
+    if (remaining_collateral_usd < 0) {
+        return (false, remaining_collateral_usd);
+    }
+    // the min collateral factor will increase as the open interest for a market increases
+    // this may lead to previously created limit increase orders not being executable
+    //
+    // the position's pnl is not factored into the remainingCollateralUsd value, since
+    // factoring in a positive pnl may allow the user to manipulate price and bypass this check
+    // it may be useful to factor in a negative pnl for this check, this can be added if required
+    let mut min_collateral_factor = market_utils::get_min_collateral_factor_for_open_interest(
+        data_store, market, values.open_interest_delta, is_long
+    );
+
+    let min_collateral_factor_for_market = market_utils::get_min_collateral_factor(
+        data_store, market.market_token
+    );
+    // use the minCollateralFactor for the market if it is larger
+    if (min_collateral_factor_for_market > min_collateral_factor) {
+        min_collateral_factor = min_collateral_factor_for_market;
+    }
+    let min_collateral_usd_for_leverage = precision::apply_factor_u128(
```

Comment:
Should call apply_factor_i128

Reply:
in gmx applyfactor uses u256 values as input and convert to int after that.  We should do the same and use `apply_factor_u128` and convert the result
`int256 maxNegativePriceImpactUsd = -Precision.applyFactor(position.sizeInUsd(), maxPriceImpactFactor).toInt256();`
https://github.com/gmx-io/gmx-synthetics/blob/77a2ff39f1414a105e8589d622bdb09ac3dd97d8/contracts/position/PositionUtils.sol#L363C13-L363C126